text/0000-introduce-a-new-signing-mechanism.md

```diff
@@ -0,0 +1,568 @@
+   ```
+
+   An issuing certificate does not change after creation.  It may therefore be cached on the gem client. Fulcio's root certificate will infrequently change.
+
+1. Validate the certificate chain in a manner similar to how [`Gem::Security::Policy`](https://github.com/rubygems/rubygems/blob/master/lib/rubygems/security/policy.rb) does it today. However, use the Rekor log entry's `integratedTime` field (seen in step 2) to determine whether the signing certificate chain was valid at the time of signing. Recall how the signing certificate is only valid for a 20 minute period.
+
+1. Extract the signature (`spec.signature.content`) from the rekord, decoding the value from base64. Using the public key included in the signing certificate, verify the signature against the gem's digest, which we compute locally.
+
+1. Rejoice!
+
+## Drawbacks
+
+We believe that the proposed approach is a sound and effective way to improve gem signing security and ease-of-use. However, the proposed approach relies on the introduction of a new service that is not controlled or operated by RubyGems (as well as relying on identity providers). This means that signing gems may be slowed if sigstore is slow, or prevented if sigstore or identity providers are unavailable. It also means that RubyGems relies on sigstore remaining financially and operationally viable in the long term.
+
+A potential alternative is for RubyGems to fully operate its own sigstore infrastructure, independently of the public instances operated by the sigstore project. We don’t think this is advisable, as it will add to the operational and financial burden on RubyGems. Additionally, there is no guarantee that a self-operated sigstore infrastructure would be any more reliable.
+
+While today sigstore is at an early stage of production readiness, we have confidence that in the long term sigstore’s infrastructure will be well funded and operated. It has funding support from the [OpenSSF](https://openssf.org/), a foundation which has [already collected 10 million dollars of funding from industry partners](https://openssf.org/press-release/2021/10/13/open-source-security-foundation-raises-10-million-in-new-commitments-to-secure-software-supply-chains/). We expect this funding to remain steady for the foreseeable future.
+
+We are less concerned about unavailability risks at verification time. This is because we believe it will be possible to cache Rekor log entries into the RubyGems bucket. Essentially, a file containing the signature and signing certificate for a .gem would live side-by-side with the .gem. Clients could retrieve the signature and certificate from the RubyGems bucket, without needing to connect to sigstore (though retaining the option to do so if sufficiently paranoid). Since installation operations greatly outnumber builds, the impact of any sigstore service problems would be contained to gem authors.
+
+## Rationale and Alternatives
```

Comment:
Was any thought given to the relatively new `ssh-keygen` signing functionality? [Git](https://calebhearth.com/sign-git-with-ssh) recently added support for signing git commits with SSH keys. We could do something similar with a new "gem" namespace for signing.

I see a few advantages to this approach over sigstore:

- Many developers will already have an SSH key used for various purposes, including access to GitHub via `git push`/`git pull`, access to remote servers, etc.
- It removes the need for a fourth party beyond the gem maintainer, the gem consumer, and RubyGems to be involved in the signing and verification process

A difficulty with ssh-keygen's signing is that it requires each gem installer to maintain an allowed_signers file with each maintainer's email, but that could be partially alleviated by RubyGems running a certificate authority signing keys as they're used to sign gems. We could explore automatically adding that CA to ~/.ssh/allowed_signers, coaching users to add it, etc.

There's also an issue of maintaining the SSH keys (at least the public), which tend to be considered disposable. I do think there would be more coaching needed to have maintainers preserve their SSH keys, but as only the public key is required for verifying we might get away with just RubyGems tracking those.

Reply:
Hey!

Sigstore maintainer here. I don't really ssh keys as orthogonal to the sigstore based ones. We can definitely add support for them in cosign, and they're already supported in rekor actually. The main issue is really the part you outlined:

> There's also an issue of maintaining the SSH keys (at least the public), which tend to be considered disposable

Rekor/Fulcio basically just automate the dispensing of disposable keys to help with this problem. We could do ssh ones or anything. We can also support people that do like to manage their own keys with the same systems.

Reply:
What if rubygems.org simply let gem owners register public keys authorized for signing? Ala github. (And let you revoke them etc, so I guess it woudl keep a record of keys that were authorized for signing during certain dates, for verifying historical gems).  That is much easier infrastructure to build than an "OIDC/OAuth IDP", I think?

Reply:
That was kind of my thought, and the CA allows them to make allowed signers maintenance easier.

Reply:
> What if rubygems.org simply let gem owners register public keys authorized for signing? 

SSH signing keys are frequently lost. If users lose track of the keys, they'll need to revoke them. We then can't trust the artifacts that were signed with that SSH key. Ephemeral keys where you trust an entity that signs the keys are going to be resistant to this.

There's very little support for SSH CAs out there. I know that Smallstep offers an [SSH CA](https://smallstep.com/docs/ssh). In my opinion, maintenance of a CA is going to be the same as maintenance of an OIDC IDP, with respect to key/cert rotation and the necessary security posture.

I also worry this would impact automation. For example, Github only issues OIDC tokens. There isn't wide library/platform support for SSH signing, we'd be relying on a very new feature.

Reply:
SSH CAs do not have a CRL function as regular X509 Certificates, making revocation a more manual process of having to add the revoked keys to a file on each client. This makes using SSH Keys as a code signing entity harder to implement (to mitigate the lack of revocation at [Keytos](https://www.keytos.io/) we create short term SSH certificates but that would not work for code signing since the packages would have to be re-signed with a new key fairly frequently. If going the certificate route I would recommend doing X509 Certificates as these CAs do have the ability to easily revoke certificates.